File: swift-bridge/Sources/SceneKitBridge/ParticleSystem.swift
import SceneKit

@_cdecl("scn_particle_system_new")
public func scn_particle_system_new() -> UnsafeMutableRawPointer? {
    scnRetain(SCNParticleSystem())
}

@_cdecl("scn_particle_system_get_birth_rate")
public func scn_particle_system_get_birth_rate(_ systemHandle: UnsafeMutableRawPointer?) -> Double {
    guard let system: SCNParticleSystem = scnBorrow(systemHandle) else { return 0 }
    return system.birthRate
}

@_cdecl("scn_particle_system_set_birth_rate")
public func scn_particle_system_set_birth_rate(_ systemHandle: UnsafeMutableRawPointer?, _ birthRate: Double) {
    guard let system: SCNParticleSystem = scnBorrow(systemHandle) else { return }
    system.birthRate = CGFloat(birthRate)
}

@_cdecl("scn_particle_system_get_life_span")
public func scn_particle_system_get_life_span(_ systemHandle: UnsafeMutableRawPointer?) -> Double {
    guard let system: SCNParticleSystem = scnBorrow(systemHandle) else { return 0 }
    return system.particleLifeSpan
}

@_cdecl("scn_particle_system_set_life_span")
public func scn_particle_system_set_life_span(_ systemHandle: UnsafeMutableRawPointer?, _ lifeSpan: Double) {
    guard let system: SCNParticleSystem = scnBorrow(systemHandle) else { return }
    system.particleLifeSpan = CGFloat(lifeSpan)
}

@_cdecl("scn_particle_system_get_loops")
public func scn_particle_system_get_loops(_ systemHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let system: SCNParticleSystem = scnBorrow(systemHandle) else { return false }
    return system.loops
}

@_cdecl("scn_particle_system_set_loops")
public func scn_particle_system_set_loops(_ systemHandle: UnsafeMutableRawPointer?, _ loops: Bool) {
    guard let system: SCNParticleSystem = scnBorrow(systemHandle) else { return }
    system.loops = loops
}

@_cdecl("scn_node_add_particle_system")
public func scn_node_add_particle_system(_ nodeHandle: UnsafeMutableRawPointer?, _ systemHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle),
          let system: SCNParticleSystem = scnBorrow(systemHandle)
    else { return }
    node.addParticleSystem(system)
}

@_cdecl("scn_node_remove_all_particle_systems")
public func scn_node_remove_all_particle_systems(_ nodeHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    node.removeAllParticleSystems()
}

@_cdecl("scn_node_particle_system_count")
public func scn_node_particle_system_count(_ nodeHandle: UnsafeMutableRawPointer?) -> Int {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return 0 }
    return node.particleSystems?.count ?? 0
}

File: src/particle_system.rs
use crate::ffi;
use crate::node::Node;
use crate::private::handle_type;

handle_type!(ParticleSystem);

impl ParticleSystem {
    /// Creates a wrapped `SCNParticleSystem` instance.
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_particle_system_new()) }
    }

    /// Mirrors `SCNParticleSystem.birthRate`.
    #[must_use]
    pub fn birth_rate(&self) -> f64 {
        unsafe { ffi::scn_particle_system_get_birth_rate(self.ptr) }
    }

    /// Sets the `SCNParticleSystem.birthRate` member.
    pub fn set_birth_rate(&self, birth_rate: f64) {
        unsafe { ffi::scn_particle_system_set_birth_rate(self.ptr, birth_rate) };
    }

    /// Mirrors `SCNParticleSystem.lifeSpan`.
    #[must_use]
    pub fn life_span(&self) -> f64 {
        unsafe { ffi::scn_particle_system_get_life_span(self.ptr) }
    }

    /// Sets the `SCNParticleSystem.lifeSpan` member.
    pub fn set_life_span(&self, life_span: f64) {
        unsafe { ffi::scn_particle_system_set_life_span(self.ptr, life_span) };
    }

    /// Mirrors `SCNParticleSystem.loops`.
    #[must_use]
    pub fn loops(&self) -> bool {
        unsafe { ffi::scn_particle_system_get_loops(self.ptr) }
    }

    /// Sets the `SCNParticleSystem.loops` member.
    pub fn set_loops(&self, loops: bool) {
        unsafe { ffi::scn_particle_system_set_loops(self.ptr, loops) };
    }
}

impl Node {
    /// Mirrors `SCNNode.addParticleSystem`.
    pub fn add_particle_system(&self, particle_system: &ParticleSystem) {
        unsafe { ffi::scn_node_add_particle_system(self.ptr, particle_system.ptr) };
    }

    /// Mirrors `SCNNode.removeAllParticleSystems`.
    pub fn remove_all_particle_systems(&self) {
        unsafe { ffi::scn_node_remove_all_particle_systems(self.ptr) };
    }

    /// Mirrors `SCNNode.particleSystemCount`.
    #[must_use]
    pub fn particle_system_count(&self) -> usize {
        unsafe { ffi::scn_node_particle_system_count(self.ptr) }
    }
}
